### ISSUE TITLE
While providing certificates, client keeps using anonymous auth

### ISSUE THREAD

**Issue by mateusz-lipowski** [kubernetes-client/csharp#319]
I'm attempting to replace token auth with certificates, but API is responding with `{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods is forbidden: User \"system:anonymous\" cannot list resource \"pods\" in API group \"\" in the namespace \"domvi-test-appservice\"","reason":"Forbidden","details":{"kind":"pods"},"code":403}`.

I'm sure that cerificate is valid and RBAC is properly configured, because kubectl works without problems and RoleBinding is the same as for token (which works properly).

I've tried multiple configurations: passing cert file paths, using raw base64 certs, reading kubeconfig that works for kubectl, all with the same result.

My current test code look like this (inspired by your AuthTests.cs):
```
var crtPath = "C:\\Users\\mateuszl\\Desktop\\cert\\domvi-test-appservice.crt";
var keyPath = "C:\\Users\\mateuszl\\Desktop\\cert\\domvi-test-appservice.key";

var config = new KubernetesClientConfiguration
{
	Host = context.Host,
	Namespace = "domvi-test-appservice",
	Username = "integration-test",
	ClientCertificateFilePath = crtPath,
	ClientKeyFilePath = keyPath,
	SkipTlsVerify = true
};

kubernetes = new Kubernetes(config);

kubernetes.ListNamespacedPod("domvi-test-appservice"); // Throws HttpOperationException
```

**Comment by brendandburns** [comment]
Can you try loading the `${KUBECONFIG}` file that works for `kubectl` using the `KuberentesClientConfiguration.BuildConfigFromConfigFile(...)` function?

If that works, then you need to figure out what is different between the code that you have above and the code that is called in that function.

If that _doesn't_ work, then please file a new issue, because we should be able to load any file that works with `kubectl`

Thanks!

**Comment by mateusz-lipowski** [comment]
Using `KubernetesClientConfiguration.BuildConfigFromConfigFile` gives exactly the same result. I guess that config loading works properly and certificate is the only issue. I also tried many other ways to load config, like putting config yaml in string and reading it or using `KubernetesClientConfiguration.BuildConfigFromConfigObject` with manually populated object.

My config is essentially copy-paste of the one generated by kubeadm, with replaced certs and namespace. That's how it looks:
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: Skipped Base64 CA Cert
    server: https://10.1.0.61:6443
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    namespace: domvi-test-appservice
    user: kubernetes-admin
  name: kubernetes-admin@kubernetes
current-context: kubernetes-admin@kubernetes
kind: Config
preferences: {}
users:
- name: kubernetes-admin
  user:
    client-certificate-data: Skipped Base64 Cert
    client-key-data: Skipped Base64 Key
```

Sould I really create a new issue? That would be just an exact copy of this one.

**Comment by mateusz-lipowski** [comment]
I did some more testing and it seems to be an issue with the way HttpClient sends certificates. According to various sources, following code should send certificate:
```
var pfxPath = "C:\\Users\\mateuszl\\Desktop\\cert\\domvi-test-appservice.pfx";

var crt = new X509Certificate2(pfxPath, "test");
var handler = new HttpClientHandler
{
	ClientCertificateOptions = ClientCertificateOption.Manual,
	SslProtocols = SslProtocols.Tls12,
	ServerCertificateCustomValidationCallback = (message, certificate2, arg3, arg4) => true,
};
handler.ClientCertificates.Add(crt);

var client = new HttpClient(handler);
var result = await client.GetAsync("https://domvi-dev-master:6443/api/v1/namespaces/domvi-test-appservice/pods?limit=500");
string content = await result.Content.ReadAsStringAsync();
```

But the result is the same as for kubernetes-client: 403 Forbidden. I think .pfx file is being loaded, because it checks password and passing an incorrenct one throws an exception. I also tried to send the request with Postman using that .pfx and URL and it worked properly.

So maybe there is something I'm missing in config or at OS level? Tried adding generated CA as trusted and installing the .pfx, but nothing changed.

**Comment by brendandburns** [comment]
Can you try it without a .pfx? (does kubectl even know how to deal with a pfx?)

The pfx conversion code in the client is likely suspect...

**Comment by mateusz-lipowski** [comment]
Using instance of `X509Certificate2` based on PEM cert has the same effect. I analysed TLS handshake for Postman and dotnet using wireshark and there are diferrences - dotnet sends empty certificate.

I also tried running the sample dotnet app on linux, the same result for both .pfx and .crt. Tried installing certificates in windows store, that also didn't help. There must be some detail in dotnet that I'm missing, it's impossible that it just doesn't work.

**Comment by mateusz-lipowski** [comment]
So I finally figured it out. It was indeed dotnet issue.

**In short**: if you use X.509 Extension `keyUsage`, you must add flag `digitalSignature`, otherwise dotnet will silently ignore the certificate.

It's absurd that the certificate will be ignored silently, especially that this flag is **not** required by kubectl and Postman. I figured it out by debugging HttpClient source code, part that caused the issue was an internal class `System.Net.Security.CertificateHelper`. Method `IsValidClientCertificate` is the one ignoring invalid certificates, it uses `IsValidForDigitalSignatureUsage` to validate `keyUsage`.

In terms of kubernetes-client project itself, it may be a good idea to add an additional validation for `keyUsage`, so it's obvious what's wrong. But then client won't be fully compliant with kubectl, since it will allow certificate without `digitalSignature` usage. Maybe there is some workaround, like mocking `keyUsage`, but I'm unsure if that would be a good idea.
